New menu bar option for i3-like functionality #844

mobile-ar opened the pull request:
New menu-bar-style configuration to add extra menu functionality ('text' monospaced default, 'squares' images indicator and 'i3' squares + not active workspaces with windows in them)

Feel free to suggest better naming for the options. I will update all needed suggestions/recomendations

Full details of this implementation in #844

danfbfern commented:
I understand that this topic is not something important for @nikitabobko, but since it still allows to have the current behaviour for users that like that simplicity, is there any constraint to prevent this from being merged? Any drawbacks of having it?
@mobile-ar keeps rebasing it to keep the topic alive for a few months

nikitabobko commented on Sources/AppBundle/TrayMenuModel.swift:
```diff
@@ -15,20 +16,52 @@ public class TrayMenuModel: ObservableObject {
+    var items: [TrayItem] = []
+    TrayMenuModel.shared.trayText = (
+        activeMode?.takeIf { $0 != mainModeId }?.first?.lets {
+            items.append(TrayItem(type: .mode, name: String($0), isActive: true))
+            return "[\($0.uppercased())] "
+        } ?? ""
+    ) +
         sortedMonitors
         .map {
-            ($0.activeWorkspace == focus.workspace && sortedMonitors.count > 1 ? "*" : "") + $0.activeWorkspace.name
+            items.append(TrayItem(type: .monitor, name: $0.activeWorkspace.name, isActive: $0.activeWorkspace == focus.workspace && sortedMonitors.count > 1))
+            return ($0.activeWorkspace == focus.workspace && sortedMonitors.count > 1 ? "*" : "") + $0.activeWorkspace.name
```
This code is written in a functional style. Please avoid side effects (`items.append`) from the lambda of `map`

mobile-ar replied:
Sure, I was just trying to reuse the iteration through the collection for performance reasons, to avoid scanning the collection twice to get the values needed for the TrayItem. I can decouple the logic from this map, but I might need to introduce a quite similar one for populating the items collection.

mobile-ar replied:
Also I use the same approach in line 23, to get the item related to the mode. I store all those inside `items`, given that in that way I have all the items to display (modes + workspaces) in an object structure instead of only text.

nikitabobko replied:
> but I might need to introduce a quite similar one for populating the items collection.

That's fine

> Also I use the same approach in line 23

Thanks for pointing out. Please fix it too.

nikitabobko commented:
I've reflected on this a bit more: Sorry, I just don't care. I hate programming UIs, I hate maintaining UIs, I hate providing configuration options for UIs – the configurations options space for UIs is enormous. For me, it's hard to come up with reasonable mental model which is powerful yet not excessive, and I just don't want to waste my mental energy on that.

I've read diagonally [the response](https://github.com/nikitabobko/AeroSpace/discussions/844#discussioncomment-11624441) from @mobile-ar and IDK, I still don't care, sorry 🤷 I appreciate the detailed response though!

The reason why I can't just merge this PR is because it touches things that I care about.
1. I care about the config format. The format should be stable, and its mental model should be clear.
2. I care about the code maintenance. It's a less of critical concern since the code in this PR is approachable and looks easily maintainable. @mobile-ar thanks for that!

**So I suggest an alternative** – please extract UI settings into a UI settings window + UserDefaults (or whatever default technology macOS apps usually use to manage settings) and don't pollute the config with UI settings. If this PR add a button like "Experimental Settings" to the dropdown, I'm ok to merge it.

Inside the window, please add a disclaimer like "These settings are EXPERIMENTAL. No stability guarantees are provided". Please also make it clear that the settings only work for macOS >= 14

Maybe, if eventually this UI logic irons out, I will be ok to make it configurable through the TOML config.

But be aware that if the UI logic turns out being too problematic to maintain, I will just drop it (hence the "experimental" in the name). I will be reasonable – I will do my minimal effort to maintain it (do the migrations if necessary), or ping @mobile-ar to ask for help (hopefully, he will be willing to help)

zhenpewu commented:
Hi,

I’ve been following this PR for a while and noticed the maintainer’s stance on UI over-customization, which I totally agree with. However, the monospaced menu bar feels excessively wide for multi-monitor workspaces. Would it be possible to provide an option to use the previous font instead of mono?

https://github.com/nikitabobko/AeroSpace/discussions/1112

mobile-ar commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -83,6 +94,87 @@ struct MonospacedText: View {
+        }
+    }
+
+    func containsEmoji() -> Bool {
```
Because an emoji inside the square images are way too small to be visible, then when someone uses emojis for the name of a workspace I just display the emoji, instead of the representation of the letter for the workspace. (Added this for support, emojis in the menu bar are quite annoying on my opinion 😄)

nikitabobko replied:
Please leave a comment in the code that they are way too small to be visible in square images

mobile-ar commented:
@nikitabobko Sure, I will be keen to move this to an "Experimental UI Settings" in the menu option and remove the TOML config for this feature. To be honest I thought you preferred all configuration for the app to be in the TOML file, that's why I implemented in that way. 

I'm totally ok with the experimental UI settings approach and save some sort of mac "UserDefauls".
The only issue might be people who likes to keep their .config files won't be able to save this config when moving between computers. (Image personal one, work one, or even on a new machine).

nikitabobko commented:
> To be honest I thought you preferred all configuration for the app to be in the TOML file, that's why I implemented in that way.

Yes, normally, I prefer plain text configuration. This case is different

> The only issue might be people who likes to keep their .config files won't be able to save this config when moving between computers

Yes, the suggested solution is a compromise

mobile-ar commented:
@nikitabobko I have updated the PR. I tried to extract the Experimental UI Settings as much as possible into the `ExperimentalUISettings.swift` in case you want to remove in the future.
I added 1 extra option to filter the empty workspaces from the menu list (because I think is was a nice to have, also was requested in comments under #711)

Looks like this now when filtered:
<img width="628" alt="image" src="https://github.com/user-attachments/assets/59140991-4ca2-4f70-8d10-5968e69799ed" />

And like this when not filtered:
<img width="629" alt="image" src="https://github.com/user-attachments/assets/0f41d04b-40dc-4c95-8d0c-b2cee758ceeb" />

mobile-ar commented:
> Inside the window, please add a disclaimer like "These settings are EXPERIMENTAL. No stability guarantees are provided". Please also make it clear in the window that the settings work only for macOS >= 14

I just added this into the setting. I forgot this part in previous commit.
Now looks like this with the requested disclaimer:

<img width="653" alt="image" src="https://github.com/user-attachments/assets/ce62dac2-33b8-4758-a0d7-378625aea1e8" />

Another option for the `menu bar style` submenu is to remove it completely putting it under `if #available(macOS 14, *)` so it won't even be shown in older versions of macOS.

manuelgcsousa commented:
Any news on this PR?

zhenpewu commented:
@mobile-ar I forked your changes https://github.com/zhenpewu/AeroSpace/actions/runs/14506543264 and publish a build artifacts for macOS 15.

cc: @manuelgcsousa

zhenpewu commented:
<img width="427" alt="Screenshot 2025-04-16 at 19 33 38" src="https://github.com/user-attachments/assets/ccf8e682-0a8d-47e3-a887-01b3b5ae1bdc" />

One nit-picky bit of feedback: the build feels a bit too large for i3-like or square layouts. I think it would be better to adjust the sizing—maybe wrap the Workspace in a square that matches the size of a monospace character. Thank you.

mobile-ar commented:
@zhenpewu both i3 style and squares are using an image representation using the native system images (you can find the reference in the SF symbols app) each square is same size as the other. These are not monospaced characters wrapped in any sort of square. 
<img width="243" alt="image" src="https://github.com/user-attachments/assets/22703248-0200-44de-9de9-fc3a1ca2d3ae" />

It's quite noticeable the empty space with numbers like 1. But I usually prefer letters, and it's less noticeable there.

<img width="364" alt="image" src="https://github.com/user-attachments/assets/1ebf7af3-fd6d-4205-8041-ffd8e72c0106" />

zhenpewu commented:
Hey!

Just to clarify—I didn’t mean the spacing inside the square, but the size of the square itself in the menu. My Swift knowledge is pretty limited, but I tried a quick tweak on my end: adding `.frame(width: 32, height: 32)` to the Image made it look a lot better. Without that, it defaults to 40, which feels a bit too big—especially since the menu bar caps at 44.

https://github.com/zhenpewu/AeroSpace/commit/d41b1eb02db78a3e8683cea9f0266d5bc7699e4d

Hope that helps!

mobile-ar commented:
@zhenpewu By Apple Design Human Interface Guideline (https://developer.apple.com/design/human-interface-guidelines/the-menu-bar#Menu-bar-extras) the menu bar extra height is 24pt (times 2 as per retina), that's why I set the height to 40 (because the image will be scaled with a factor of 2) that will make the image 20pt. I feel more than enough padding while maintaining good readability inside the square.

nikitabobko commented on Sources/AppBundle/ExperimentalUISettings.swift:
```diff
@@ -0,0 +1,62 @@
+}
+
+@MainActor
+func getExperimentalUISettingsMenu(viewModel: TrayMenuModel) -> some View {
```
UX problem: users don't know how the tray is going to look like without clicking it. But when they click it, the menu disappears, which is not nice

I see several ways to fix it:
1. Change the tray icon, when the mouse is hovered over the item (it may introduce tray icon and menu jaggering or not)
2. Don't hide the menu when user clicks on items (it can also potentially may introduce tray icon and menu jaggering)
3. Extract this submenu into a real NSWindow
4. Maybe something else...

You are free to choose yourself

mobile-ar replied:
So I went with something else 😄 (given that not closing the menu on selection looks not possible in SwiftUI as per my investigation) to keep it simple and to use the same UI to display how the menu bar will look in real time (reusing the MenuLabel).
Looks like this:

<img width="670" alt="image" src="https://github.com/user-attachments/assets/f29d9e3b-79f0-4656-b13b-5718db870b30" />

nikitabobko replied:
Thanks, that's also a nice option. Now the problem is that when it takes 2 rows, it looks too bulky, it'd be nice to merge 2 rows into a single row

mobile-ar replied:
Menu is not giving a lot of options to make it nicer. I will prefer a spacer between the image and the text and align one to leading and other to trailing but it's not possible.
This is the best looking 1 line I could think of: 
<img width="648" alt="image" src="https://github.com/user-attachments/assets/c2139b1a-3663-4d6f-9767-a50371e3cbca" />

nikitabobko commented on Sources/AppBundle/ExperimentalUISettings.swift:
```diff
@@ -0,0 +1,62 @@
+            UserDefaults.standard.synchronize()
+        }
+    }
+    var filterEmptyWorkspacesFromMenu: Bool {
```
Please drop this config option for now. I plan to address the problem with `persistent-workspaces` config option: https://github.com/nikitabobko/AeroSpace/issues/999

mobile-ar replied:
Removed.

nikitabobko commented on Sources/AppBundle/TrayMenuModel.swift:
```diff
@@ -25,10 +27,42 @@ public class TrayMenuModel: ObservableObject {
+    let isActive: Bool
+    var systemImageName: String {
+        let lowercasedName = name.lowercased()
+        switch type {
+            case .mode:
+                return "\(lowercasedName).circle"
+            case .monitor:
+                if isActive {
+                    return "\(lowercasedName).square.fill"
+                } else {
+                    return "\(lowercasedName).square"
+                }
+        }
```
This is fragile

mobile-ar replied:
True that. I improved this by using this only for the charset supported with system images (1 to 9 and A to Z). So this is now optional and if nil the UI is layout in a custom way.

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -92,6 +110,96 @@ struct MonospacedText: View {
+                "Monospaced font"
+            case .systemText:
+                "System font"
+            case .squares:
+                "Square images"
```
The problem with this style is that the tray icon disappears if users do smth like this: `aerospace workspace dev`

mobile-ar replied:
Fixed / improved how long name workspaces are displayed for `squares` and `i3` styles

nikitabobko reviewed (commented):
Sorry, I don't have enough time lately to review all the PRs, and I didn't take a thorough look at the code yet, but here are my comments so far

I also somehow managed to make the tray icon completely disappear when running your code. Can you please make it clear how to clear UserDefaults settings from the CLI. I tried something like `defaults delete bobko.aerospace.debug` but it didn't work.

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -92,6 +110,96 @@ struct MonospacedText: View {
+}
+
+extension String {
+    func containsEmoji() -> Bool {
```
```suggestion
    // If workspace name contains emojis we use the plain emoji in text to avoid visibility issues scaling the emoji to fit the squares
    fileprivate func containsEmoji() -> Bool {
```

mobile-ar replied:
Fixed

mobile-ar commented:
> Sorry, I don't have enough time lately to review all the PRs, and I didn't take a thorough look at the code yet, but here are my comments so far
> 
> I also somehow managed to make the tray icon completely disappear when running your code. Can you please make it clear how to clear UserDefaults settings from the CLI. I tried something like `defaults delete bobko.aerospace.debug` but it didn't work.

You can delete it using `defaults delete AeroSpaceApp` for the debug one. (Remember that you need to trigger some change for the menu to refresh the value, UserDefaults it's not an observable object)

nikitabobko commented on Sources/AppBundle/ExperimentalUISettings.swift:
```diff
@@ -0,0 +1,68 @@
+    let appearence = NSApplication.shared.effectiveAppearance.name
+    let color = (appearence == .vibrantDark || appearence == .darkAqua) ? Color.white : Color.black
+    return Menu {
+        Text("These settings are EXPERIMENTAL.")
+        Text("No stability guarantees are provided.")
+        Divider()
```
```suggestion
```

nikitabobko replied:
For some reason if I drop these lines, I get a segmentation fault if I run aerospace like `./run-debug.sh`

nikitabobko replied:
Actually, if I do almost any changes to this file, then the next time I run the debug build from CLI, I get a segmentation fault.

I have learnt that I can resolve the segmentation fault via `rm -rf .build/arm64-apple-macosx`, but it's not a nice dev experience. I understand that it's probably a swift bug, but we have to come up with some workaround (rewriting code in some way to avoid these incremental compilation bugs would be sufficient)

I have the following swift version installed via swiftly

```
Apple Swift version 6.1 (swift-6.1-RELEASE)
Target: arm64-apple-macosx15.0
```

I observe the same problem with other Swift setups:
- If I run debug build from CLI using swift 6.1 from Xcode
- If I run debug build from Xcode itself

nikitabobko replied:
Example of a crash:

```
💣 Program crashed: Bad pointer dereference at 0x0000000000000009

Thread 0 crashed:

 0 0x0000000196e74fdc swift_retain + 24 in libswiftCore.dylib
 1 0x0000000196ea36cc swift::OpaqueValue* tuple_initializeWithCopy<false, false>(swift::OpaqueValue*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*) + 104 in libswiftCore.dylib
 2 0x00000001ebe40b34 <deduplicated_symbol> + 64 in SwiftUICore
 3 0x00000001b590c618 initializeWithCopy for Menu + 156 in SwiftUI
 4 menuBar(viewModel:) + 136 in AeroSpaceApp at /Users/bobko/a/AeroSpace/Sources/AppBundle/MenuBar.swift:7:5

     5│ @MainActor
     6│ public func menuBar(viewModel: TrayMenuModel) -> some Scene { // todo should it be converted to "SwiftUI struct"?
     7│     MenuBarExtra {
      │     ▲
     8│         let shortIdentification = "\(aeroSpaceAppName) v\(aeroSpaceAppVersion) \(gitShortHash)"
     9│         let identification      = "\(aeroSpaceAppName) v\(aeroSpaceAppVersion) \(gitHash)"

 5 AeroSpaceApp.body.getter + 228 in AeroSpaceApp at /Users/bobko/a/AeroSpace/Sources/AeroSpaceApp/AeroSpaceApp.swift:18:9

    16│     @MainActor // macOS 13
    17│     var body: some Scene {
    18│         menuBar(viewModel: viewModel)
      │         ▲
    19│     }
    20│ }

 6 0x00000001ec12c6e8 BodyAccessor.setBody(_:) + 1204 in SwiftUICore
 7 0x00000001b668ef70 AppBodyAccessor.updateBody(of:changed:) + 152 in SwiftUI
 8 0x00000001ec12edd8 closure #1 in DynamicBody.updateValue() + 600 in SwiftUICore
 9 0x00000001ec12e804 DynamicBody.updateValue() + 928 in SwiftUICore
10 0x00000001b6920728 AG::Graph::UpdateStack::update() + 524 in AttributeGraph
11 0x00000001b6920f94 AG::Graph::update_attribute(AG::data::ptr<AG::Node>, unsigned int) + 420 in AttributeGraph
12 0x00000001b692911c AG::Graph::input_value_ref_slow(AG::data::ptr<AG::Node>, AG::AttributeID, unsigned int, unsigned int, AGSwiftMetadata const*, unsigned char&, long) + 712 in AttributeGraph
13 0x00000001b693faf8 AGGraphGetValue + 228 in AttributeGraph
14 0x00000001ec12e714 DynamicBody.updateValue() + 688 in SwiftUICore
15 0x00000001b6920728 AG::Graph::UpdateStack::update() + 524 in AttributeGraph
16 0x00000001b6920f94 AG::Graph::update_attribute(AG::data::ptr<AG::Node>, unsigned int) + 420 in AttributeGraph
17 0x00000001b692911c AG::Graph::input_value_ref_slow(AG::data::ptr<AG::Node>, AG::AttributeID, unsigned int, unsigned int, AGSwiftMetadata const*, unsigned char&, long) + 712 in AttributeGraph
18 0x00000001b693faf8 AGGraphGetValue + 228 in AttributeGraph
19 0x00000001b53ffe7c ResolvedMenuBarExtra.MakeList.updateValue() + 548 in SwiftUI
20 0x00000001b53d7528 specialized implicit closure #1 in closure #1 in closure #1 in Attribute.init<A>(_:) + 24 in SwiftUI
21 0x00000001b6920728 AG::Graph::UpdateStack::update() + 524 in AttributeGraph
22 0x00000001b6920f94 AG::Graph::update_attribute(AG::data::ptr<AG::Node>, unsigned int) + 420 in AttributeGraph
23 0x00000001b69287b0 AG::Graph::value_ref(AG::AttributeID, unsigned int, AGSwiftMetadata const*, unsigned char&) + 272 in AttributeGraph
24 0x00000001b693fb48 AGGraphGetValue + 308 in AttributeGraph
25 0x00000001b562a444 AppGraph.filteredPrimarySceneSummaries.getter + 3496 in SwiftUI
26 0x00000001b668e76c AppGraph.instantiateOutputs() + 480 in SwiftUI
27 0x00000001ec303b34 GraphHost.instantiate() + 224 in SwiftUICore
28 0x00000001b569ce88 closure #1 in closure #1 in runApp<A>(_:) + 88 in SwiftUI
29 0x00000001ec132d28 closure #1 in static Update.ensure<A>(_:) + 56 in SwiftUICore
30 0x00000001ec131b04 static Update.ensure<A>(_:) + 96 in SwiftUICore
31 0x00000001b569ce1c runApp<A>(_:) + 124 in SwiftUI
32 0x00000001b59fcd14 static App.main() + 224 in SwiftUI

Backtrace took 1.50s

/Users/bobko/a/AeroSpace/run-debug.sh: line 6: 74831 Segmentation fault: 11  ./.debug/AeroSpaceApp "$@"
```

mobile-ar replied:
I'm not able to reproduce this issue. I can edit this file and never get any issues. I also run this this trough the CLI all the time.
<img width="593" alt="image" src="https://github.com/user-attachments/assets/7c2d0337-0fbc-45b6-b69c-ebfed4fa9902" />


I have also swift 6.1
<img width="548" alt="image" src="https://github.com/user-attachments/assets/7d0a6cf0-730a-4ae2-b5d9-6f876edbd98e" />

I will try to do some more investigation to see if I can reproduce this.

mobile-ar replied:
@nikitabobko  I tried adding changes to that same file with Zed and vim, but still running normally for me with `./run-debug.sh`. 
I have installed Xcode 16.2 but I'm actively using Xcode 16.3.
I can't reproduce this issue. Also it's super weird issue for a fairly common file, there is nothing complex on it to produce something like this as per my understanding.

nikitabobko commented on Sources/AppBundle/ExperimentalUISettings.swift:
```diff
@@ -0,0 +1,68 @@
+            }
+        }
+    } label: {
+        Text("Experimental UI Settings")
```
```suggestion
        Text("Experimental UI Settings (No stability guarantees)")
```

mobile-ar replied:
Fixed

nikitabobko commented on Sources/AppBundle/ExperimentalUISettings.swift:
```diff
@@ -0,0 +1,68 @@
+                MenuBarLabel(viewModel.trayText, textStyle: .system, color: color)
+            }
+        }
+        Text(MenuBarStyle.monospacedText.title)
+        Button {
+            viewModel.experimentalUISettings.displayStyle = .monospacedText
+        } label: {
+            Toggle(isOn: .constant(viewModel.experimentalUISettings.displayStyle == .monospacedText)) {
+                MenuBarLabel(viewModel.trayText, color: color)
+            }
+        }
```
I think monospaced text should be the first as its the default

mobile-ar replied:
Done

nikitabobko commented on Sources/AppBundle/TrayMenuModel.swift:
```diff
@@ -25,10 +27,51 @@ public class TrayMenuModel: ObservableObject {
+
+enum TrayItemType: String, Hashable {
+    case mode
+    case monitor
```
Why do you call it `monitor` if it's for workspaces?

Anyway, I think better names for these two would be: `circle` and `square`

nikitabobko replied:
Ok, given this line: `if item.type == .mode`, I see how `mode` & `workspace` names are better

nikitabobko replied:
But I still have a question why `monitor` and not `workspace` name

mobile-ar replied:
Fixed

nikitabobko commented:
Can you please everything related to UI to `Sources/AppBundle/ui`?

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -26,6 +26,8 @@ public func menuBar(viewModel: TrayMenuModel) -> some Scene { // todo should it
             }
```
IDK if it's something you are interested, but I'd be happy to refactor this `func menuBar` function to a more canonical Swift UI code (in a separate PR). I know nothing about Swift UI and I just wrote it as a function because it's the way people program in Jetpack Compose (Android reactive UI framework, which I am happen to be familiar with)

But I have a feeling that a canonical way to split UI into a components is via structs in Swift UI. I have a feeling that you might know Swift UI better

nikitabobko replied:
The same about `func getExperimentalUISettingsMenu` & `MenuBarLabel.itemView` which you introduce in this PR

mobile-ar replied:
I have seen both implementations in SwiftUI both functions and structs, usually is common to use functions when the view receives parameters. 

Sure I will be interested in doing the refactoring later in another PR. I think I tried that in one of my views and I was having some weird behavior (some changes not refreshing the UI) but I could take a look at this.

mobile-ar replied:
Yes, I don't recall any difference between both. Could be reusability at some point depending the scope of the functions, but it's not the case here. 

I think there is some small gain in performance from the function version, but it might be minimal.
Some source: https://www.swiftbysundell.com/articles/swiftui-views-versus-modifiers/

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -62,25 +64,51 @@ public func menuBar(viewModel: TrayMenuModel) -> some Scene { // todo should it
     } label: {
         if viewModel.isEnabled {
-            MonospacedText(viewModel.trayText)
+            switch viewModel.experimentalUISettings.displayStyle {
+                case .systemText:
+                    MenuBarLabel(viewModel.trayText, textStyle: .system)
+                case .monospacedText:
+                    MenuBarLabel(viewModel.trayText)
+                case .squares:
+                    MenuBarLabel(viewModel.trayText, trayItems: viewModel.trayItems)
+                case .i3:
+                    MenuBarLabel(viewModel.trayText, trayItems: viewModel.trayItems, workspaces: viewModel.workspaces)
+            }
```
Right now, you unconditionally pass `viewModel.trayText` even if it's not used inside (for square and i3 styles)

I feel like pairs of systemText/monospacedText styles and pairs of squares/i3 styles are significantly different from each other to dedicate separate `MenuBarLabel`s for the two groups

mobile-ar replied:
TrayText is always used as a fallback for the imageRenderer in case of failure and as a label for the image. It's called text here:
`Image(cgImage, scale: 2, label: Text(text))` line 114 (body of `MenuBarLabel`)

Maybe should I rename the parameter to be `trayText` to better reflect this?

nikitabobko replied:
Makes sense, thanks

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -62,25 +64,51 @@ public func menuBar(viewModel: TrayMenuModel) -> some Scene { // todo should it
-struct MonospacedText: View {
+@MainActor
+struct MenuBarLabel: View {
     @Environment(\.colorScheme) var colorScheme: ColorScheme
```
Those tricks with `@Environment(\.colorScheme)` and `NSApplication.shared.effectiveAppearance` inside `getExperimentalUISettingsMenu` is something that makes me think if it's the the way it should be done. They are related yet apparently they mean different things

Do you have any pointers?

mobile-ar replied:
Yes, this was quite tricky, I couldn't find a lot of documentation. But when using the `MenuBarLabel` inside the menu, the color of the colorScheme was not reflecting the change of color.

There is one color for System -> Appearance (`NSApplication.shared.effectiveAppearance`)
<img width="475" alt="image" src="https://github.com/user-attachments/assets/e358433d-08f7-45ea-bd3b-78a4ca60b4c7" />

And other for System -> Wallpaper
<img width="489" alt="image" src="https://github.com/user-attachments/assets/c4769bf7-409e-462a-a6be-d778dc81d1c2" />

When you start combining this 2 macOS system configurations `@Environment(\.colorScheme)` is not getting the correct value inside the menu. But when combining `colorScheme` and `effectiveAppearance`) it started working for all the combinations.

nikitabobko replied:
Thank you, so it wasn't some guide on the internet, but you figured it out with series of trials and errors.

Do I get it correctly that you have figured what combinations of `@Environment(\.colorScheme)` and `NSApplication.shared.effectiveAppearance` correspond to which Light/Dark Appearance and Light/Dark Wallpaper?

If so, I'd suggest to refactor the code to make it self-documenting (if Apple doesn't make a nice API, we will do it ourselves):

```swift
// System Settings -> Appearance -> Light/Dark
enum AppearanceTheme {
  case light
  case dark

  static var current: AppearanceTheme {
    // @Environment(\.colorScheme) and NSApplication.shared.effectiveAppearance magic...
  }
}

// System Settings -> Wallpaper -> Light/Dark
enum WallpaperTheme {
  case light
  case dark

  static var current: WallpaperTheme {
    // @Environment(\.colorScheme) and NSApplication.shared.effectiveAppearance magic...
  }
}
```

Would smth like this make sense? Would this API be easier to use?

mobile-ar replied:
I don't think this is possible, `@Environment` is a `DynamicProperty`
```swift
/// An interface for a stored variable that updates an external property of a
/// view.
///
/// The view gives values to these properties prior to recomputing the view's
/// ``View/body-swift.property``.
@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
public protocol DynamicProperty {
```
And also from `@Environment` init documentation
```swift
...
///     struct MyView: View {
///         @Environment(\.colorScheme) var colorScheme: ColorScheme
///
///         // ...
///     }
///
/// SwiftUI automatically updates any parts of `MyView` that depend on
/// the property when the associated environment value changes.
...
```
Also even if we try to access `EnvironmentValues` manually, we will only get defaults values, and not the live updates
```swift
@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
public struct EnvironmentValues : CustomStringConvertible {

/// Creates an environment values instance.
///
/// You don't typically create an instance of ``EnvironmentValues``
/// directly. Doing so would provide access only to default values that
/// don't update based on system settings or device characteristics.
/// Instead, you rely on an environment values' instance
/// that SwiftUI manages for you when you use the ``Environment``
/// property wrapper and the ``View/environment(_:_:)`` view modifier.
public init()
```

I don't think this will be possible without some hack like having an empty view that uses this property and hook on change of `colorScheme` to update a `@Published` property inside the our custom `Appearance: ObservableObject`

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -92,6 +120,129 @@ struct MonospacedText: View {
+                                .bold()
+                        }
+                    }
+                    if workspaces != nil {
```
You pass `workspaces` array but you don't use it in `MenuBarLabel`

mobile-ar replied:
It's used to check if I need to display the workspaces and as a trick redraw the view on changes when the workspace gets a window added.
What I noticed is that I can add 1 extra property `isEffectivelyEmpty` to `WorkspaceViewModel` and then use workspaces instead of `Workspace.all.filter ...`

nikitabobko replied:
>  What I noticed is that I can add 1 extra property isEffectivelyEmpty to WorkspaceViewModel and then use workspaces instead of Workspace.all.filter ...

Yes, that would be better

nikitabobko replied:
>  What I noticed is that I can add 1 extra property isEffectivelyEmpty to WorkspaceViewModel and then use workspaces instead of Workspace.all.filter ...

That would be better, thanks

mobile-ar replied:
Fixed

nikitabobko commented:
Two digit workspaces sometimes don't look nice (on my screenshot, it's fine for square style, but not ok for i3 style)

<img width="280" alt="image" src="https://github.com/user-attachments/assets/3f7f1433-71ef-4b29-8940-ca26c5f1dd17" />

nikitabobko commented on Sources/AppBundle/MenuBar.swift:
```diff
@@ -92,6 +120,129 @@ struct MonospacedText: View {
+                        }
+                    }
+                    if workspaces != nil {
+                        let otherWorkspaces = Workspace.all.filter { workspace in
+                            !workspace.isEffectivelyEmpty && !trayItems.contains(where: { item in item.type == .monitor && item.name == workspace.name })
```
Realization. It's unnecessary O(n^2):
```swift
/*...*/.filter {
  /*...*/ && !trayItems.contains(/*...*/)
}
```

Please avoid it. I think it will be fixed automatically once this comment is resolved https://github.com/nikitabobko/AeroSpace/pull/1049/files#r2059296197

mobile-ar replied:
I still need this to filter the active monitors workspaces, to not display it duplicated ones. Maybe I need to filter this in some other way that I'm not aware.
I could not filter workspaces inside TrayMenuModel, because they are used in the main menu to display all workspaces.

Also trayItems array should be small (macOS only supports 5 monitors in it best hardware) + 1 mode <- (Not a justification, I will try to think how to improve this)

mobile-ar replied:
Fixed by adding isVisible to the WorkspaceViewModel

mobile-ar commented:
> Two digit workspaces sometimes don't look nice (on my screenshot, it's fine for square style, but not ok for i3 style)
> 
> Especially, given that `dev` workspace looks nice
> 
> <img alt="image" width="280" src="https://private-user-images.githubusercontent.com/20517828/437225379-3f7f1433-71ef-4b29-8940-ca26c5f1dd17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU2NjU5NzcsIm5iZiI6MTc0NTY2NTY3NywicGF0aCI6Ii8yMDUxNzgyOC80MzcyMjUzNzktM2Y3ZjE0MzMtNzFlZi00YjI5LTg5NDAtY2EyNmM1ZjFkZDE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI2VDExMDc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3MjA4ZDNmNjY1MjNmOWI0YjBkZTQ4MzQxNjI0YWZkNDE2ZTUyNTdlMTU2NjBjMTFmMDRmNzBkNzMyYjZjNDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.A-mSVNuXisr7xQGQGm6ox06U5QdWTYCElxSn0PBAdO4">

I added a fix for this. There is support for numbers in system images until 50. I added that fix. Still not sure why this 2 digits numbers did not display correctly. I will investigate this more.

nikitabobko commented:
@mobile-ar thanks for your work and patience! Merged

mobile-ar commented:
> @mobile-ar thanks for your work and patience! Merged

All good 😄 I'm happy to contribute. Thank you for creating and maintaining this awesome tool.

mobile-ar commented:
> I didn't understand what the complication around `appearanceObserver` is for, so I cleaned up your code [9f49504](https://github.com/nikitabobko/AeroSpace/commit/9f49504049d0bc90826e87a33ff82db877284483)
> 
> Please let me know if it was important for some reason

It's important if you want to avoid an small case where the UI color inside the menu is incorrect.

If you open System Sesttings -> Appearance -> and you change the current Appearance. And without any other screen refresh or windows change you open the Aerospace menu -> Experimental UI, you will see the color is the previous one. 

It will show incorrectly like this:

<img width="1032" alt="image" src="https://github.com/user-attachments/assets/aab3217c-fc3b-49cb-b93c-3bcd1e5fd312" />

That's why I made Appearance to be observable, to avoid this small issue. By making it observable, you ensure that on change of Appearance, the UI is refreshed correctly every time.

nikitabobko commented:
@mobile-ar thanks for the explanation. I think the appearance doesn't change that often, and I prefer to trade correctness for more straightforward code in this place.

It's unfortunate that Apple makes us write this boilerplate to avoid this bug.

volker-fr commented:
I want to express my disappointment that this PR was closed.

Would it be possible to have it as "permanently experimental" with the notion of "if it breaks, there will be zero support"?


In my humble opinion, this feature would resolve a major pain point that was worked around by various people in often not so nice or buggy ways. The sole reason why I have to use SketchyBar is due to the lack of seeing the workspace -- something this PR would solve. And Sketchybar has its own issues / conflicts with the official macOS bar.

I understand that this implementation is a bit hacky and not as smooth as hoped. At the same time, my impression is that the way AeroSpace has to work in a few areas to get around Apples limitation. One example here is requirement for [Proper monitor arrangement](https://nikitabobko.github.io/AeroSpace/guide#proper-monitor-arrangement). I personal would prefer a little bit more inconvenience here, over the lack of having this big missing feature.


In the end the decision is up to @nikitabobko , I still think that this feature would bring tremendous value.

zhenpewu commented:
@volker-fr this PR is closed, becasue @nikitabobko have merged it in his end.
